fix: Use attributes from the active PG connection

cbandy opened the pull request:
As discovered and addressed in #142, the `pg` gem and its underlying `libpq` library allow multiple hosts and network addresses in the connection string. The solution applied there, however, does not always return accurate information about the connected peer.

When there are multiple hosts or addresses, the library [tries each in order until one succeeds](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-MULTIPLE-HOSTS). The [`target_session_attrs`](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNECT-TARGET-SESSION-ATTRS) setting allows both primary and replica servers to be specified in the connection string, and the library will figure which is which. Always using the first value in the connection string ignores this stuff.

I don't see an explanation for why `conninfo_hash` was used in 463c14c3581641ce57270c1c7cfb91715cfd2c3e, so perhaps I'm missing some benefit to it.

The separate methods (`host`, `hostaddr`, `port`, `user`, and `db`) describe the singular active connection from the resolved connection string (IPv4, IPv6, primary, replica, etc). I'm pretty sure we should use these when they're available.

There are a couple of older versions of `libpq` that allow multiple hosts/addresses but lack a `hostaddr` method. For those, `conninfo_hash` is accurate when it returns a single address. When it returns multiple addresses, we can use the first like #142 or omit `net.peer.ip`. The latter seems better to me, so that's what I've done in this PR.

~TODO:~
- [x] I would like to rebase this after #178 lands for the UNIX socket test there.

cbandy commented on instrumentation/pg/lib/opentelemetry/instrumentation/pg/patches/connection.rb:
```diff
@@ -123,38 +123,38 @@ def generated_postgres_regex
-                'net.peer.ip' => conninfo_hash[:hostaddr]&.to_s,
-                'net.peer.port' => first_in_list(conninfo_hash[:port]&.to_s)
+                'net.peer.ip' => transport_addr&.to_s,
+                'net.peer.port' => port&.to_s
```
🤔 I'm curious about this line. Semantic Conventions say that `peer.port` should be an integer. Is `to_s` appropriate here?

- https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/span-general/

ahayworth replied:
I don't actually recall what the reasons were for stringification here; which suggests that perhaps there was no valid reason and it was just a mistake. If you want to include a fix for this in the PR, that would be a welcome change. 😄

ahayworth commented:
This looks like a good change, I will happily review it once you're ready. I agree with everything outlined in the PR description, including the `TODO` you added. Thank you so much for this!

---

> I don't see an explanation for why conninfo_hash was used in https://github.com/open-telemetry/opentelemetry-ruby-contrib/commit/463c14c3581641ce57270c1c7cfb91715cfd2c3e, so perhaps I'm missing some benefit to it.

For context, this was contributed before we split off the `-contrib` repo, so the original PR that added this instrumentation is here: https://github.com/open-telemetry/opentelemetry-ruby/pull/659

I don't recall having any specific reason for using `conninfo_hash`; if there are better ways to get that information than we should use it. I think it was just my best attempt at getting the right information since I am not an expert on the `pg` gem itself. We are most grateful for contributions that improve the quality of instrumentation in this way - we are certainly not experts and we're covering a very wide surface area in this repo. 😍

cbandy commented:
I haven't forgotten about this! `let(:port) { nil }` exposed a bug in the `pg` gem I'm waiting on now: https://github.com/ged/ruby-pg/issues/491

The implementation here will consider versions of that gem with and without whatever fix comes.

ahayworth commented:
@cbandy looks like they just merged a fix, too!

cbandy commented:
Rebased and ready to go. I've added a test verifying that active connection information is returned when using "multiple hosts."

IP address and port depend on the version of the gem and the underlying driver. The original implementation was correct to use `conninfo_hash` in the past. Recent versions have the desired behavior, so this detects and uses those.

I made three other changes to the transport attributes after reading the [semantic conventions](https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/span-general/):

1. `net.peer.port` is now an integer.
2. `net.transport` is now `ip_tcp` when not using a Unix socket.
3. `net.transport` is _not set_ when using a Unix socket. Instead, `net.sock.family` is set to `unix`.

I wanted to fill out `net.sock.peer.addr` for the Unix socket case, but there isn't a good way to get that information from the gem or driver. I left `net.peer.name` set to the _directory_ containing the socket, as we've done before.

cbandy commented:
Amended for Rubocop.

ahayworth commented on instrumentation/pg/lib/opentelemetry/instrumentation/pg/patches/connection.rb:
```diff
@@ -123,43 +123,56 @@ def generated_postgres_regex
+          def transport_addr
+            # The hostaddr method is available when the gem is built against
+            # a recent version of libpq.
+            return hostaddr if defined?(hostaddr)
```
Huh. I actually didn't know you could use `defined?` in that way (I thought this was maybe calling `defined?(result-of-unintentional-hostaddr-call)` which wouldn't be a good thing to rely on). But wouldn't you know it, it's right there in the docs 🤦 

https://ruby-doc.org/docs/keywords/1.9/Object.html#method-i-defined-3F

Thank you, I learned something new today. 😄

cbandy replied:
I learned it during this PR as well. I saw it as part of module prepending: `super if defined? super`

In this case, we can use `respond_to?`. I wonder if one is definitely faster? 🤷🏻

ahayworth replied:
I mean; either is probably fine. But my first meeting of the morning cancelled so that means its time for [MicroBenchmark, Thanksgiving Edition 🦃](https://gist.github.com/ahayworth/bdd5f12c5ce08a075289bfa4530a468c)

Looks like `defined?` may be a tiny bit faster? It probably doesn't have to go through things like `method_missing` and whatever. Here are the results:
```
/-------------------------------------------------------------\
|                  user     system      total        real     |
|    defined?:   0.094393   0.000956   0.095349 (  0.095374)  |
| respond_to?:   0.099900   0.001003   0.100903 (  0.100944)  |
\-------------------------------------------------------------/
   \
    \                                  ,+*^^*+___+++_
     \                           ,*^^^^              )
      \                       _+*                     ^**+_
       \                    +^       _ _++*+_+++_,         )
                _+^^*+_    (     ,+*^ ^          \+_        )
               {       )  (    ,(    ,_+--+--,      ^)      ^\
              { (@)    } f   ,(  ,+-^ __*_*_  ^^\_   ^\       )
             {:;-/    (_+*-+^^^^^+*+*<_ _++_)_    )    )      /
            ( /  (    (        ,___    ^*+_+* )   <    <      \
             U _/     )    *--<  ) ^\-----++__)   )    )       )
              (      )  _(^)^^))  )  )\^^^^^))^*+/    /       /
            (      /  (_))_^)) )  )  ))^^^^^))^^^)__/     +^^
           (     ,/    (^))^))  )  ) ))^^^^^^^))^^)       _)
            *+__+*       (_))^)  ) ) ))^^^^^^))^^^^^)____*^
            \             \_)^)_)) ))^^^^^^^^^^))^^^^)
             (_             ^\__^^^^^^^^^^^^))^^^^^^^)
               ^\___            ^\__^^^^^^))^^^^^^^^)\\
                    ^^^^^\uuu/^^\uuu/^^^^\^\^\^\^\^\^\^\
                       ___) >____) >___   ^\_\_\_\_\_\_\)
                      ^^^//\\_^^//\\_^       ^(\_\_\_\)
                        ^^^ ^^ ^^^ ^
```

ahayworth commented on instrumentation/pg/lib/opentelemetry/instrumentation/pg/patches/connection.rb:
```diff
@@ -123,43 +123,56 @@ def generated_postgres_regex
-              'db.user' => conninfo_hash[:user]&.to_s,
-              'db.name' => database_name,
-              'net.peer.name' => first_in_list(conninfo_hash[:host]&.to_s)
+              'db.user' => user,
+              'db.name' => db
```
I'm guessing these get defined in the `pg` connection?

cbandy replied:
Yes. These are methods of ::PG:: Connection. I don't mind using `self.db` to make that more clear. Would that be better?

ahayworth replied:
No need to change, this one was actually just a question. Thank you!

ahayworth commented on instrumentation/pg/lib/opentelemetry/instrumentation/pg/patches/connection.rb:
```diff
@@ -123,43 +123,56 @@ def generated_postgres_regex
           def transport_attrs
-            if conninfo_hash[:host]&.start_with?('/')
-              { 'net.transport' => 'Unix' }
+            host = self.host
+            if host&.start_with?('/')
+              {
+                'net.sock.family' => 'unix',
+                'net.peer.name' => host
+              }
```
This kind of scope-shenanigans makes my eye twitch, as I try to parse exactly which `host` we're actually looking at (the variable? the method call?). Even if it is nicer to read!

Would you mind doing something silly like `the_host` (or whatever feels most readable to you) so that we aren't shadowing the host method in this scope here? I know it's a minor thing, but...

cbandy replied:
Agreed. I wasn't sure what to name it either. Is `h` too short?

ahayworth replied:
`h` is fine with me 😄

ahayworth reviewed (commented):
No major issues, and I'm grateful that this PR is so thorough. And because you've been so thorough and helpful I have rewarded you with a little finicky comment to address before we merge it. 😆 ❤️  Great job catching the semantic convention changes, as well!

As a meta question, should we test multiple versions of the PG gem via the `Appraisals` file? I don't think I have a strong opinion either way.

ahayworth commented on instrumentation/pg/lib/opentelemetry/instrumentation/pg/patches/connection.rb:
```diff
@@ -123,43 +123,56 @@ def generated_postgres_regex
-              'db.user' => conninfo_hash[:user]&.to_s,
-              'db.name' => database_name,
-              'net.peer.name' => first_in_list(conninfo_hash[:host]&.to_s)
+              'db.user' => user,
+              'db.name' => db
```
@ericmustin I think what would happen is that we would end up checking the attribute's validity [here](https://github.com/open-telemetry/opentelemetry-ruby/blob/a4efaf3a44b0efc6ca805a6d0a5bdc4f13a01b34/sdk/lib/opentelemetry/sdk/internal.rb#L27), and that doesn't like nil-things. The net effect when you follow things around the codebase is that we would log a warning, but ultimately just [keep only the valid values](https://github.com/open-telemetry/opentelemetry-ruby/blob/a4efaf3a44b0efc6ca805a6d0a5bdc4f13a01b34/sdk/lib/opentelemetry/sdk/trace/span.rb#L341-L345).

You're correct that `nil` would be correctly represented as the empty string when the destination is supposed to be an OTLP `AnyValue`, but I think the actual controlling part of the spec for attribute values is [here](https://opentelemetry.io/docs/reference/specification/common/#attribute) ... and that explicitly mentions that `null` (`nil`, in our  case) is actually not allowed as an attribute value. `AnyValue` is used [more restrictively](https://opentelemetry.io/docs/reference/specification/common/attribute-type-mapping/) from what I can gather - at the collector, and in the logging SDK?

But regardless, we shouldn't let `nil` sneak through and I'm really glad you brought it up. @cbandy can we assert with any amount of confidence that the `user` and `host` methods never return `nil`? If they can (even if only in corner cases), we should make sure to `reject` away `nil` values.

ericmustin replied:
i truly do not know C but i think looking at the c extension, nil is a possible return value https://github.com/ged/ruby-pg/blob/cf1443dd4473ffc0b1ebbe401c6a700b1cd50476/ext/pg_connection.c#L615-L634 so i think we should coerce to empty string to be safe

cbandy commented:
> As a meta question, should we test multiple versions of the PG gem via the `Appraisals` file? I don't think I have a strong opinion either way.

I'll try that out. I can use the [RubyGems dataset](https://play.honeycomb.io/ruby-together/datasets/rubygems.org/) at Honeycomb to pick some useful versions.

cbandy commented:
Amended `host` variable and added appraisals for some interesting `pg` versions:

```
>> BUNDLE_GEMFILE=/app/instrumentation/pg/gemfiles/pg_1.2.gemfile bundle list pg
/bundle/gems/pg-1.2.3
>> BUNDLE_GEMFILE=/app/instrumentation/pg/gemfiles/pg_1.3.gemfile bundle list pg
/bundle/gems/pg-1.3.5
>> BUNDLE_GEMFILE=/app/instrumentation/pg/gemfiles/pg_latest.gemfile bundle list pg
/bundle/gems/pg-1.4.5
```

ahayworth commented:
@cbandy Thanks so much for the back-and-forth on this. One more ✅ and I think we're good to merge. (cc @ericmustin since you've been looking)

ericmustin commented:
kicked the ci, i think you have to pull in https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/211 for things to pass. :shakes-fist-at-jruby:
